File: problems/滑动窗口/239.滑动窗口最大值.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    deque<int> dq;
    vector<int> result;
    for (int i = 0; i < nums.size(); i++) {
        while (!dq.empty() && nums[dq.back()] <= nums[i])
            dq.pop_back();
        dq.push_back(i);
        if (dq.front() <= i - k)
            dq.pop_front();
        if (i >= k - 1)
            result.push_back(nums[dq.front()]);
    }
    return result;
}

int main() {
    // 测试用例
    vector<int> n1 = {1, 3, -1, -3, 5, 3, 6, 7};
    auto r1 = maxSlidingWindow(n1, 3);
    for (int x : r1) cout << x << " ";
    cout << endl;  // 预期: 3 3 5 5 6 7

    vector<int> n2 = {1};
    auto r2 = maxSlidingWindow(n2, 1);
    for (int x : r2) cout << x << " ";
    cout << endl;  // 预期: 1

    vector<int> n3 = {1, -1};
    auto r3 = maxSlidingWindow(n3, 1);
    for (int x : r3) cout << x << " ";
    cout << endl;  // 预期: 1 -1

    return 0;
}

File: problems/滑动窗口/239.滑动窗口最大值.py
#!/usr/bin/env python3
"""
239. 滑动窗口最大值
算法: 单调递减双端队列
时间复杂度: O(n)
空间复杂度: O(k)
"""
from collections import deque


def max_sliding_window(nums: list[int], k: int) -> list[int]:
    deque_indices = deque()
    result = []

    for i in range(len(nums)):
        while deque_indices and nums[deque_indices[-1]] <= nums[i]:
            deque_indices.pop()
        deque_indices.append(i)

        if deque_indices[0] <= i - k:
            deque_indices.popleft()

        if i >= k - 1:
            result.append(nums[deque_indices[0]])
    return result


if __name__ == "__main__":
    print(max_sliding_window([1, 3, -1, -3, 5, 3, 6, 7], 3))
    # 预期: [3, 3, 5, 5, 6, 7]
    print(max_sliding_window([1], 1))  # 预期: [1]
    print(max_sliding_window([1, -1], 1))  # 预期: [1, -1]
